### PR TITLE
Certificate table

### PR DESCRIPTION
**Pull request by snhenson**
##### Checklist

- [x] tests are added or updated

This PR adds a certificate table containing properties of each certificate type: specifically the NID of the public key algorithm and the auth mask associated with it. This is used to remove a lot of the algorithm specific cases and make it easier to add new types in future.

It also fixes a bug in the signature selection code and adds a test for that case and some others.

### CODE REVIEW

File: ssl/ssl_cert.c [hunk 1]
```diff
@@ -976,3 +976,41 @@ int ssl_ctx_security(const SSL_CTX *ctx, int op, int bits, int nid, void *other)
     return ctx->cert->sec_cb(NULL, ctx, op, bits, nid, other,
                              ctx->cert->sec_ex);
 }
+
+/*
+ * Certificate table information. NB: table entries must match SSL_PKEY indices
```
Comment (richsalz) [inline ssl/ssl_cert.c]: should we add a test that enforces this?  Like the cipher-table test @ekasper did?

File: ssl/statem/statem_clnt.c [hunk 2]
```diff
@@ -3334,62 +3329,51 @@ int tls_construct_client_certificate(SSL *s, WPACKET *pkt)
...
-            /* check failed */
-            SSLerr(SSL_F_SSL3_CHECK_CERT_AND_ALGORITHM, SSL_R_BAD_ECC_CERT);
-            goto f_err;
+    clu = ssl_cert_lookup_by_pkey(X509_get0_pubkey(s->session->peer), &idx);
+
+    /* Check certificate is recognised and suitable for cipher */
+    if (clu == NULL || (alg_a & clu->amask) == 0) {
+        int err;
+
+        if (alg_a & SSL_aECDSA) {
```
Comment (richsalz) [inline ssl/statem/statem_clnt.c]: are all three error codes worth it?  or just "missing signing cert" ?  side-note.

Comment (snhenson) [inline ssl/statem/statem_clnt.c]: I kept them for compatibility with the previous version. I've no problem with replacing them with a single case: it simplifies it and removes more algorithm specific code if I do.

File: ssl/ssl_cert.c (hunk 1 above)
Comment (snhenson) [inline ssl/ssl_cert.c]: I'll have a look how we might do that... 

An assert would do it to, well maybe not ;-)

**Comment by snhenson** [comment]
Update: rebased comments addressed and additional test added.

File: ssl/ssl_cert_table.h [hunk 3]
```diff
@@ -0,0 +1,21 @@
+/*
+ * Copyright 2015-2016 The OpenSSL Project Authors. All Rights Reserved.
```
Comment (FdaSilvaYY) [inline ssl/ssl_cert_table.h]: 2017 ?

File: ssl/t1_lib.c [hunk 4]
```diff
@@ -2299,6 +2264,22 @@ int ssl_security_cert_chain(SSL *s, STACK_OF(X509) *sk, X509 *x, int vfy)
...
 /*
+ * For TLS 1.2 servers check if we have a certificate which can be used
+ * with the signature algorithm "lu".
+ */
+
+static int tls12_check_cert_sigalg(const SSL *s, const SIGALG_LOOKUP *lu)
+{
+    const SSL_CERT_LOOKUP *clu = ssl_cert_lookup_by_idx(lu->sig_idx);
+
+    /* If not reconised or not supported by cipher mask it is not suitable */
```
Comment (FdaSilvaYY) [inline ssl/t1_lib.c]: Typo 'reconised'

File: ssl/ssl_cert_table.h (hunk 3 above)
Comment (snhenson) [inline ssl/ssl_cert_table.h]: Fixed

File: ssl/t1_lib.c (hunk 4 above)
Comment (snhenson) [inline ssl/t1_lib.c]: Fixed

**Comment by snhenson** [comment]
Any more comments?

**Comment by levitte** [comment]
Since there was VMS related stuff in the internal test program, I let my VMS environment have a go.  Have no fear, it passed with flying colours.

**Comment by snhenson** [comment]
Pushed, thanks everyone for reviews and comments.

**Pull request closed without merging**